scripts/create-darwin-volume.sh

```diff
@@ -1,33 +1,210 @@
+    # (~diskXsY)--but I'm retaining the more-semantic approach since
+    # it documents intent better.
+    # /usr/sbin/diskutil info -plist / | xmllint --xpath "/plist/dict/key[text()='ParentWholeDisk']/following-sibling::string[1]/text()" -
+    #
+    local special_device
+    special_device="$(/usr/bin/stat -f "%Sd" /)"
+    echo "${special_device/s[0-9]*/}"
 }
 
-find_nix_volume() {
-    diskutil apfs list -plist "$1" | xmllint --xpath "(/plist/dict/array/dict/key[text()='Volumes']/following-sibling::array/dict/key[text()='Name']/following-sibling::string[starts-with(translate(text(),'N','n'),'nix')]/text())[1]" - 2>/dev/null || true
+# Strongly assuming we'll make a volume on the device root is on
+# But you can override NIX_VOLUME_USE_DISK to create it on some other device
+readonly NIX_VOLUME_USE_DISK="${NIX_VOLUME_USE_DISK:-$(root_disk_identifier)}"
+
+substep() {
+    printf "   %s\n" "" "- $1" "" "${@:2}"
+}
+
+# $1 = label
+volumes_labeled() {
+    xsltproc --novalid --stringparam label "$1" - <(ioreg -ra -c "AppleAPFSVolume") <<'EOF'
+<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
+  <xsl:output method="text"/>
+  <xsl:template match="/">
+    <xsl:apply-templates select="/plist/array/dict/key[text()='IORegistryEntryName']/following-sibling::*[1][text()=$label]/.."/>
+  </xsl:template>
+  <xsl:template match="dict">
+    <xsl:apply-templates match="string" select="key[text()='BSD Name']/following-sibling::*[1]"/>
+    <xsl:text>◊</xsl:text>
+    <xsl:apply-templates match="string" select="key[text()='UUID']/following-sibling::*[1]"/>
+    <xsl:text>&#xA;</xsl:text>
+  </xsl:template>
+</xsl:stylesheet>
+EOF
+    # I cut label out of the extracted values, but here it is for reference:
+    # <xsl:apply-templates match="string" select="key[text()='IORegistryEntryName']/following-sibling::*[1]"/>
+    # <xsl:text>◊</xsl:text>
+}
+
+# $1 = volume special (i.e., disk1s7)
+right_disk() {
+    [ "${1::${#NIX_VOLUME_USE_DISK}}" = "$NIX_VOLUME_USE_DISK" ]
+}
+# $1 = volume special (i.e., disk1s7)
+right_volume() {
+    # if set, it must match; otherwise ensure it's on the right disk
+    if [ -z "$NIX_VOLUME_USE_SPECIAL" ]; then
+        if right_disk "$1"; then
+            NIX_VOLUME_USE_SPECIAL="$1" # latch on
+            return 0
+        else
+            return 1
+        fi
+    else
+        [ "$1" = "$NIX_VOLUME_USE_SPECIAL" ]
+    fi
+}
+# $1 = volume uuid
+right_uuid() {
+    # if set, it must match; otherwise allow
+    if [ -z "$NIX_VOLUME_USE_UUID" ]; then
+        NIX_VOLUME_USE_UUID="$1" # latch on
+        return 0
+    else
+        [ "$1" = "$NIX_VOLUME_USE_UUID" ]
+    fi
+}
+
+# $1 = label
+cure_volumes() {
+    local found="" volume special uuid
+    # loop just in case they have more than one volume
+    # (nothing stops you from doing this)
+    for volume in $(volumes_labeled "$NIX_VOLUME_LABEL"); do
+        IFS=◊ read -r special uuid <<< "$volume"
```

Comment:
Why are we using this unicode character as a separator? Neither UUIDs nor disk BSD names will have spaces, so we can just use a space, or maybe use a `=` instead.

If we did have to worry about spaces then the `for` loop wouldn’t work (as it does word splitting, not line splitting). And in fact I’m tempted to say we should replace it anyway for idiomatic reasons. If we replace the separator with `=` then we can turn this loop into

```bash
    while IFS== read -r special uuid; do
        # ...
    done <<<"$(volumes_labeled "$NIX_VOLUME_LABEL")"
```

Reply:
Vestigial from when this was still handling volume labels (before I decided not to try to support both `Nix` and `Nix Store`). Another good catch! 👍 Adopted locally.

Reply:
> If we replace the separator with `=` then we can turn this loop into
> 
> ```shell
>     while IFS== read -r special uuid; do
>         # ...
>     done <<<"$(volumes_labeled "$NIX_VOLUME_LABEL")"
> ```

Didn't think to detail the for loop when I triaged this before, but I'm auditing suggestions to make sure I got them all and realized I probably should (I'll also add a comment on this...)

The for loop decision was driven by a gotcha lurking in the function calls, here. The right paths through `cure_volume` can require user interactions that are also being done with `read`, and passing on stdin at the outer level interferes with them (causing them to skip).

You should be able to see this with something like:
```bash
heh(){
    set -x
    while read -r blah; do
        echo "blah? $blah"
        while read -r -p "What do you think?" heh; do
            echo "Do we get here? $heh"
        done
    done <<< "blah blah"
    set +x
}
heh
```

I'm not sure off the top of my head if there's a better fix that doesn't make the code harder to understand. (I noticed `read -u 1` or `read -u 2` "work" here, but I don't grok why and haven't tried to figure it out.)

Reply:
`read -u 1` works because stdout and stderr by default point to the terminal (just like stdin) and therefore can actually be read from. A better workaround there is to write it like `read -r -p "What do you think?" heh </dev/tty`, as that's explicitly pointing `read` at the current terminal. Though it would be safe to duplicate fd 0 instead, in case the script is running with stdin pointed elsewhere. We could also update the confirm routines to use a duplicated fd explicitly so we don't have to worry about them.

Having said that, we could also just use fd 3 ourselves and leave fd 0 alone. That would look like

```bash
while IFS== read -r -u 3 special uuid; do
  # ...
done 3<<<"$(volumes_labeled "$NIX_VOLUME_LABEL")"
```

Of course, this assumes fd 3 is free to use, but AFAIK we're not using it anywhere else in the install scripts.